### ISSUE TITLE
Carbon dialog not centering properly when it's parent is a positioned element.

### ISSUE THREAD

**Issue by piotr-szachewicz** [Sage/carbon#328]
If the carbon dialog is placed in an DOM element that is positioned (e.g. somewhere in the `<div id="ui-main">` element in `sage_one_advanced`), the dialog is positioned depending on that element not on the  whole window.

the code for centering the dialog in Carbon: 

``` js
  centerDialog = () => {
    let height = this._dialog.offsetHeight / 2,
        width = this._dialog.offsetWidth / 2,
        midPointY = window.innerHeight / 2 + window.pageYOffset,
        midPointX = window.innerWidth / 2 + window.pageXOffset;

    midPointY = midPointY - height;
    midPointX = midPointX - width;

    if (midPointY < 20) {
      midPointY = 20;
    } else if (Bowser.ios) {
      midPointY -= window.pageYOffset;
    }

    if (midPointX < 20) {
      midPointX = 20;
    }

    this._dialog.style.top = midPointY + "px";
    this._dialog.style.left = midPointX + "px";
  }
// https://github.com/Sage/carbon/blob/v0.21.0/src/components/dialog/dialog.js#L135-L156
```

The code for centering the dialog in Sage One legacy UI:

``` coffeescript
  center: ($element, zeroScrollTop) ->
    elementHeight = $element.outerHeight()
    elementWidth = $element.outerWidth()
    windowHeight = window.height()
    windowWidth = window.width()
    scrollTop = window.scrollTop()
    scrollLeft = window.scrollLeft()
    calculatedTop = ((windowHeight - elementHeight) / 2) + scrollTop

    if (zeroScrollTop)
      # Still want window.scrollTop to be calculated initially, even if we remove it later.
      calculatedTop = calculatedTop - scrollTop

    if (calculatedTop < 50)
      calculatedTop = 50

    $element.offset
      top: calculatedTop,
      left: ((windowWidth - elementWidth) / 2) + scrollLeft
```

I tried to calculate midpoints like this:

```
midPointX = this.refs.dialog.offsetParent.offsetWidth / 2
midPointY = this.refs.dialog.offsetParent.offsetHeight / 2
```

which works well for my dialog, but makes the dialog to be rendered a bit too high.

Probably we should use a similar code to the one used in legacy Sage One, because it was already tested in a lot of contexts.

**Comment by toni-leigh** [comment]
How does this do now the new changes to Dialog are on their way?

**Comment by andrewjtait** [comment]
Looking further into the solution suggested on https://github.com/Sage/carbon/issues/328#issuecomment-193706378

There is already a library which does this for React to resolve the same issues called react-portal https://www.npmjs.com/package/react-portal

We should maybe look to implement this for things such as Modals, Tooltips, Error Messages, Date Pickers.

**Comment by andrewjtait** [comment]
Fixed as of v2.2.0
